src/main/groovy/lang/GroovyClassLoader.java

```diff
@@ -969,13 +970,24 @@ public Void run() {
      */
     public void clearCache() {
         synchronized (classCache) {
+            for (Class cl : classCache.values()) {
+                InvokerHelper.removeClass(cl);
+            }
             classCache.clear();
         }
         synchronized (sourceCache) {
             sourceCache.clear();
         }
     }
 
+    @Override
+    public void close() throws IOException {
+        super.close();
+        for (Class cl : classCache.values()) {
```

Comment:
I thought that synchronization might not be an issue since we are `close`-ing the ClassLoader anyway and I wanted to avoid the additional cost of synchronizing over and clearing `sourceCache`. But I guess it's better safe than sorry, so I made the change.